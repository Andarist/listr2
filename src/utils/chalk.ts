import chalkDefault from 'chalk'

let chalk: chalkDefault.Chalk

/* istanbul ignore if */
if (process.env?.LISTR_DISABLE_COLOR !== '1' && chalkDefault.supportsColor) {
  chalk = new chalkDefault.Instance()
} else {
  chalk = new chalkDefault.Instance({ level: 0 })
}

/** A custom chalk instance to disable coloring compeletely. */
export default chalk
