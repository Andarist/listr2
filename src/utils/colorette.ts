import { createColors } from 'colorette'

export default createColors({ useColor: process.env?.LISTR_DISABLE_COLOR !== '1' })
