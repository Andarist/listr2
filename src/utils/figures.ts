import { isUnicodeSupported } from './is-unicode-supported'

const FIGURES_MAIN = {
  warning: '⚠',
  cross: '✖',
  arrowDown: '↓',
  tick: '✔',
  arrowRight: '→',
  pointer: '❯',
  checkboxOn: '☒',
  arrowLeft: '←',
  squareSmallFilled: '◼',
  pointerSmall: '›'
}

const FIGURES_FALLBACK = {
  ...FIGURES_MAIN,
  warning: '‼',
  cross: '×',
  tick: '√',
  pointer: '>',
  checkboxOn: '[×]',
  squareSmallFilled: '■'
}

export const figures = isUnicodeSupported() ? FIGURES_MAIN : FIGURES_FALLBACK
