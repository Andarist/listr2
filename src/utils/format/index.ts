export * from './colorette'
export * from './indent-string'
export * from './figures'
export * from './parse-time'
export * from './splat'
export * from './timestamp'
