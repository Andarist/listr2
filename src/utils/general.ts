import * as clone from 'clone'

/**
 * Deep clones a object in the most easiest manner.
 */
export function cloneObject<T extends Record<PropertyKey, any>> (obj: T): T {
  return clone(obj)
}
