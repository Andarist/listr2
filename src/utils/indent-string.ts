export function indentString (string: string, count: number): string {
  return string.replace(/^(?!\s*$)/gm, ' '.repeat(count))
}