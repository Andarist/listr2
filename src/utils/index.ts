export * from './environment'
export * from './format'
export * from './logger'
export * from './process-output'
export * from './ui'
export * from './assert'
export * from './clone'
export * from './concurrency'
export * from './delay'
export * from './uuid'
