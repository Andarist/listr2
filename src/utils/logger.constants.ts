/** Default loglevels for the logger */
export enum LogLevels {
  silent = 'silent',
  fail = 'fail',
  skip = 'skip',
  success = 'success',
  data = 'data',
  start = 'start',
  title = 'title',
  retry = 'retry',
  rollback = 'rollback'
}
