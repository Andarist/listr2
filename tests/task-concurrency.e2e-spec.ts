import type { MockProcessOutput } from '@tests/utils'
import { expectProcessOutputToMatchSnapshot, delay, mockProcessOutput, unmockProcessOutput } from '@tests/utils'

import type { ListrTask } from '@root'
import { Listr } from '@root'

describe('concurrency', () => {
  const output: MockProcessOutput = {} as MockProcessOutput

  beforeEach(async () => {
    mockProcessOutput(output)
  })

  afterEach(async () => {
    unmockProcessOutput(output)
    jest.clearAllMocks()
  })

  const tasks: ListrTask<any, any>[] = [
    {
      title: '1',
      task: async (): Promise<void> => {
        await delay(35)
      }
    },

    {
      title: '2',
      task: async (): Promise<void> => {
        await delay(19)
      }
    },

    {
      title: '3',
      task: async (): Promise<void> => {
        await delay(9)
      }
    },

    {
      title: '4',
      task: async (): Promise<void> => {
        await delay(0)
      }
    }
  ]

  it('should run tasks in parallel', async () => {
    await new Listr(tasks, { concurrent: true, renderer: 'test' }).run()

    expectProcessOutputToMatchSnapshot(output, 'WKPv1yKZBa2LY7IStPAL9emHNpaTSsQQ')
  })

  it.each([ 1, 2, 3, 4 ])('should limit the concurrency: %d', async (concurrent) => {
    await new Listr(tasks, { concurrent, renderer: 'test' }).run()
  })

  it('should limit the concurrency to 2', async () => {
    await new Listr(tasks, { concurrent: 2, renderer: 'test' }).run()

    expectProcessOutputToMatchSnapshot(output, '2pZ6rwMSxvmHpaVE3btDMTePhnvK0a3s')
  })

  it('should run tasks sequentially', async () => {
    await new Listr(tasks, { concurrent: false, renderer: 'test' }).run()

    expectProcessOutputToMatchSnapshot(output, 'wqpFCLvXmbjnc3DNDBL76AWAK8uxEHlo')
  })
})
