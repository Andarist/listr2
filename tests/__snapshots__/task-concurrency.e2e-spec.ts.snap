// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`concurrency should limit the concurrency to 2: 2pZ6rwMSxvmHpaVE3btDMTePhnvK0a3s-exit 1`] = `[]`;

exports[`concurrency should limit the concurrency to 2: 2pZ6rwMSxvmHpaVE3btDMTePhnvK0a3s-stderr 1`] = `[]`;

exports[`concurrency should limit the concurrency to 2: 2pZ6rwMSxvmHpaVE3btDMTePhnvK0a3s-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"title":"1","hasFinalized":false}}
",
  "{"event":"STATE","data":"STARTED","task":{"title":"2","hasFinalized":false}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"title":"2","hasFinalized":true}}
",
  "{"event":"STATE","data":"STARTED","task":{"title":"3","hasFinalized":false}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"title":"3","hasFinalized":true}}
",
  "{"event":"STATE","data":"STARTED","task":{"title":"4","hasFinalized":false}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"title":"4","hasFinalized":true}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"title":"1","hasFinalized":true}}
",
]
`;

exports[`concurrency should run tasks in parallel: WKPv1yKZBa2LY7IStPAL9emHNpaTSsQQ-exit 1`] = `[]`;

exports[`concurrency should run tasks in parallel: WKPv1yKZBa2LY7IStPAL9emHNpaTSsQQ-stderr 1`] = `[]`;

exports[`concurrency should run tasks in parallel: WKPv1yKZBa2LY7IStPAL9emHNpaTSsQQ-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"title":"1","hasFinalized":false}}
",
  "{"event":"STATE","data":"STARTED","task":{"title":"2","hasFinalized":false}}
",
  "{"event":"STATE","data":"STARTED","task":{"title":"3","hasFinalized":false}}
",
  "{"event":"STATE","data":"STARTED","task":{"title":"4","hasFinalized":false}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"title":"4","hasFinalized":true}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"title":"3","hasFinalized":true}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"title":"2","hasFinalized":true}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"title":"1","hasFinalized":true}}
",
]
`;

exports[`concurrency should run tasks sequentially: wqpFCLvXmbjnc3DNDBL76AWAK8uxEHlo-exit 1`] = `[]`;

exports[`concurrency should run tasks sequentially: wqpFCLvXmbjnc3DNDBL76AWAK8uxEHlo-stderr 1`] = `[]`;

exports[`concurrency should run tasks sequentially: wqpFCLvXmbjnc3DNDBL76AWAK8uxEHlo-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"title":"1","hasFinalized":false}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"title":"1","hasFinalized":true}}
",
  "{"event":"STATE","data":"STARTED","task":{"title":"2","hasFinalized":false}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"title":"2","hasFinalized":true}}
",
  "{"event":"STATE","data":"STARTED","task":{"title":"3","hasFinalized":false}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"title":"3","hasFinalized":true}}
",
  "{"event":"STATE","data":"STARTED","task":{"title":"4","hasFinalized":false}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"title":"4","hasFinalized":true}}
",
]
`;
